src/virtualdevice.cc

```diff
@@ -0,0 +1,111 @@
+// This file is part of Projecteur - https://github.com/jahnf/projecteur - See LICENSE.md and README.md
+#include "virtualdevice.h"
+
+#include <fcntl.h>
+
+#include <QDebug>
+
+class VirtualDevice;
+
+VirtualDevice::VirtualDevice() {
+  setupVirtualDevice();
+}
+
+VirtualDevice::~VirtualDevice() {
+  if (deviceCreated) {
+    ioctl(uinp_fd, UI_DEV_DESTROY);
+    close(uinp_fd);
+    qDebug("uinput Device Closed");
+  }
+}
+
+// Setup uinput device that can send mouse and keyboard events. Logs the result too.
+void VirtualDevice::setupVirtualDevice() {
+  int i=0;
+  if (deviceCreated)
+    return;
+
+  // Open the input device
+  uinp_fd = open("/dev/uinput", O_WRONLY | O_NDELAY);
+  if (uinp_fd < 0) {
+    qDebug("Unable to open /dev/uinput\n");
+    deviceCreated = false;
+    return;
+  }
+
+  memset(&uinp,0,sizeof(uinp));
+  // Intialize the UINPUT device to NULL
+  strncpy(uinp.name, "Projecteur_input_device", UINPUT_MAX_NAME_SIZE);
+  uinp.id.bustype = BUS_USB;
+
+  // Setup the uinput device
+  ioctl(uinp_fd, UI_SET_EVBIT, EV_KEY);
+  ioctl(uinp_fd, UI_SET_EVBIT, EV_REL);
+  ioctl(uinp_fd, UI_SET_RELBIT, REL_X);
+  ioctl(uinp_fd, UI_SET_RELBIT, REL_Y);
+
+  for (i=0; i < 256; i++) {
+    ioctl(uinp_fd, UI_SET_KEYBIT, i); }
+
+  ioctl(uinp_fd, UI_SET_KEYBIT, BTN_MOUSE);
+  ioctl(uinp_fd, UI_SET_KEYBIT, BTN_TOUCH);
+
+  ioctl(uinp_fd, UI_SET_KEYBIT, BTN_MOUSE);
+  ioctl(uinp_fd, UI_SET_KEYBIT, BTN_LEFT);
+  ioctl(uinp_fd, UI_SET_KEYBIT, BTN_MIDDLE);
+  ioctl(uinp_fd, UI_SET_KEYBIT, BTN_RIGHT);
+  ioctl(uinp_fd, UI_SET_KEYBIT, BTN_FORWARD);
+  ioctl(uinp_fd, UI_SET_KEYBIT, BTN_BACK);
+
+  // Create input device into input sub-system
+  write(uinp_fd, &uinp, sizeof(uinp));
+  if (ioctl(uinp_fd, UI_DEV_CREATE)) {
+    qDebug("Unable to create Virtual (UINPUT) device.");
+    deviceCreated = false;
+    return;
+  }
```

Comment:
Close uinp_fd, otherwise it will not get closed.
Alternative: check in constructor if `uinp_fd >= 0` and close